### PR TITLE
fix(4669/4671/4696/4697): template notebook generated from bucket, must also include retentionRules.

### LINKED ISSUE #4669: Cant explore  the newly created buckets
## About the bug

**Steps to reproduce:**
List the minimal actions needed to reproduce the behavior.

1. Try to explore any created bucket then the following message pop up. Preloaded buckets works fine. 
image

2. ...
3. ...

**Expected behavior:**

**Actual behavior:**

**Visual Proof:**

## About your environment

**Environment info:**

**Config:**

**Logs:**

**Performance:**
Generate profiles with the following commands for bugs related to performance, locking, out of memory (OOM), etc.

```sh
# Commands should be run when the bug is actively.
# Note: This command will run for at least 30 seconds.
curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=true"
curl -o vars.txt "http://localhost:8086/debug/vars"
iostat -xd 1 30 > iostat.txt
# Attach the `profiles.tar.gz`, `vars.txt`, and `iostat.txt` output files.
```

### LINKED ISSUE #4671: InfluxDB Cloud - Cannot explore buckets in Chrome
## About the bug

**Steps to reproduce:**
List the minimal actions needed to reproduce the behavior.

1. This only happens when I am using Chrome that is signed in to the same Google account that I use to sign in to InfluxDB Cloud
2. When I go to the Load Data > Buckets page I can view all buckets
3. Any bucket that I click on I am unable to explore, including the system buckets

**Expected behavior:**
I expect to be able to explore a bucket when I click on it, I tried in Firefox and in an Incognito window and it works fine.
 
**Actual behavior:**
The page does not render and it says "An InfluxDB error has occurred please report the issue here"

**Visual Proof:**

https://user-images.githubusercontent.com/42261599/170300083-d00d8cd1-bc95-40a9-8f30-2a6449478c4d.mov

## About your environment

I'm using a free InfluxDB Cloud on Chrome Version 101.0.4951.64 (Official Build) (arm64) with a MacBook Air (M1, 2020) 

**Config:**

**Logs:**

**Performance:**
Generate profiles with the following commands for bugs related to performance, locking, out of memory (OOM), etc.

```sh
# Commands should be run when the bug is actively.
# Note: This command will run for at least 30 seconds.
curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=true"
curl -o vars.txt "http://localhost:8086/debug/vars"
iostat -xd 1 30 > iostat.txt
# Attach the `profiles.tar.gz`, `vars.txt`, and `iostat.txt` output files.
```

### LINKED ISSUE #4696: "An InfluxDB error has occurred" when I just click the bucket name on Load Data > Buckets.
## About the bug

**Steps to reproduce:**
List the minimal actions needed to reproduce the behavior.

1. ...
2. ...
3. ...

**Expected behavior:**

**Actual behavior:**

**Visual Proof:**

## About your environment

**Environment info:**

**Config:**

**Logs:**

**Performance:**
Generate profiles with the following commands for bugs related to performance, locking, out of memory (OOM), etc.

```sh
# Commands should be run when the bug is actively.
# Note: This command will run for at least 30 seconds.
curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=true"
curl -o vars.txt "http://localhost:8086/debug/vars"
iostat -xd 1 30 > iostat.txt
# Attach the `profiles.tar.gz`, `vars.txt`, and `iostat.txt` output files.
```

### LINKED ISSUE #4697: An influxDB error has occurred when I was trying to load the data bucket
## About the bug

**Steps to reproduce:**
List the minimal actions needed to reproduce the behavior.

1. ...
2. ...
3. ...

**Expected behavior:**

**Actual behavior:**

**Visual Proof:**

## About your environment

**Environment info:**

**Config:**

**Logs:**

**Performance:**
Generate profiles with the following commands for bugs related to performance, locking, out of memory (OOM), etc.

```sh
# Commands should be run when the bug is actively.
# Note: This command will run for at least 30 seconds.
curl -o profiles.tar.gz "http://localhost:8086/debug/pprof/all?cpu=true"
curl -o vars.txt "http://localhost:8086/debug/vars"
iostat -xd 1 30 > iostat.txt
# Attach the `profiles.tar.gz`, `vars.txt`, and `iostat.txt` output files.
```

### PR DESCRIPTION
**Pull request by wiedld**
Closes #4669 
Closes #4671 
Closes #4696 
Closes #4697 

## Debugging data shown:
* steps as identified by Will Baker.
   1. make new bucket
   2. click on new bucket name
   3. --> it tries to generate a template notebook (when `influx.set('exploreWithFlows’, true)` as in prod)
   4. black screen of doom.
        * local ts-node dev server: `TypeError: bucket.retentionRules is undefined.` 
        * transpiled js code error: Error for `find()` on the expected array.
* Issue:
  * Was generating a template notebook with only the passed in name (and a hard-coded type=='user'). Then failing on an expectation of retention values in getReadableRetention().
  * fix: provide template array. Page now loads without error.
* Issue:
  * BucketContext is out of sync with the generated new bucket in the reducer entities.
  * This is because the TEMPLATE.init uses the passed init structure, and gives preference to this init structure.
     * the init structure is only black templating items. Not the bucket crud object. 
  * fix: fetch the actual bucket object
     * handle if only have bucket name (since `bucket.id` is optional prop in openapi...presumably for older/legacy buckets).

### CODE REVIEW

**Comment by wiedld** [comment]
> Looks good to me so far, what's the TODO fix part about in the comments?

The notebook generated (and crud object persisted to the backend) uses this bucket-on-init from the notebooks template. Not the full bucket (entity object) which is already persisted in the redux state.

So we get the bucket `{name: "DLW 2", type: "user"}` persisted within the created notebook:
Screen Shot 2022-05-31 at 6 47 34 PM

File: src/buckets/components/BucketCard.tsx [hunk 1]
```diff
@@ -38,7 +38,9 @@ const BucketCard: FC<Props & RouteComponentProps<{orgID: string}>> = ({
 }) => {
   const handleNameClick = () => {
     if (isFlagEnabled('exploreWithFlows')) {
-      history.push(`/${PROJECT_NAME.toLowerCase()}/from/bucket/${bucket.name}`)
+      history.push(
+        `/${PROJECT_NAME.toLowerCase()}/from/bucket/${bucket.name}/${bucket.id}`
```
[the remark this answers is not in the document]
Comment (rockstar) [reply src/buckets/components/BucketCard.tsx:42]: Why do we namespace this across `name` _and_ `id`?

**Review by rockstar — APPROVED** [review]
One small question, but certainly not anything that should (necessarily) hold up merge.

**Comment by wiedld** [comment]
> One small question, but certainly not anything that should (necessarily) hold up merge.

The bucket type definitions:
* come from the openapi contract, which generates the types under `src/client/` when we run yarn generate.
* The contracted definition has the id as an optional parameter, and the name as a required parameter.
   * (I am guessing this is due to legacy buckets, or other scenarios, without the id. But either way -- we need to support.)
* The template-generating code (`src/flows/templates/types/bucket.ts`) was updated in this PR:
   * to treat the backend bucket crud object as the src of truth
   * attempt to fetch by id first.
   * on failure (a.k.a. no id), then fetch all buckets and match on name.

The previous/old code (now changed) was:
* generating a "bucket" object with only a name and a few init (hardcoded default) props.
* was creating the initial persisted notebook crud object, containing this init bucket.

Now we only generate a notebook crud object, with a real bucket attached.

**Pull request merged**